src/xtea.cpp

```diff
@@ -0,0 +1,213 @@
+        data[j+5] = static_cast<uint8_t>(right[i] >> 8u);
+        data[j+6] = static_cast<uint8_t>(right[i] >> 16u);
+        data[j+7] = static_cast<uint8_t>(right[i] >> 24u);
+    }
+}
+
+template<size_t BLOCK_SIZE>
+void XTEA_decrypt(uint8_t data[BLOCK_SIZE * 8], const key& k)
+{
+    alignas(16) uint32_t left[BLOCK_SIZE], right[BLOCK_SIZE];
+    for (auto i = 0u, j = 0u; i < BLOCK_SIZE; i += 1u, j += 8u) {
+        left[i] = data[j] | data[j+1] << 8u | data[j+2] << 16u | data[j+3] << 24u;
+        right[i] = data[j+4] | data[j+5] << 8u | data[j+6] << 16u | data[j+7] << 24u;
+    }
+
+    uint32_t sum = delta * 32;
+    for (auto i = 0u; i < 32; ++i) {
+        for (auto j = 0u; j < BLOCK_SIZE; ++j) {
+            right[j] -= (((left[j] << 4) ^ (left[j] >> 5)) + left[j]) ^ (sum + k[(sum >> 11) & 3]);
+        }
+        sum -= delta;
+        for (auto j = 0u; j < BLOCK_SIZE; ++j) {
+            left[j] -= (((right[j] << 4) ^ (right[j] >> 5)) + right[j]) ^ (sum + k[(sum) & 3]);
+        }
+    }
+
+    for (auto i = 0u, j = 0u; i < BLOCK_SIZE; i += 1u, j += 8u) {
+        data[j] = static_cast<uint8_t>(left[i]);
+        data[j+1] = static_cast<uint8_t>(left[i] >> 8u);
+        data[j+2] = static_cast<uint8_t>(left[i] >> 16u);
+        data[j+3] = static_cast<uint8_t>(left[i] >> 24u);
+        data[j+4] = static_cast<uint8_t>(right[i]);
+        data[j+5] = static_cast<uint8_t>(right[i] >> 8u);
+        data[j+6] = static_cast<uint8_t>(right[i] >> 16u);
+        data[j+7] = static_cast<uint8_t>(right[i] >> 24u);
+    }
+}
+
+} // anonymous namespace
+
+void encrypt(uint8_t* data, size_t length, const key& k)
+{
+    assert(length % 8 == 0);
+
+#ifdef __AVX512F__
+    {
+        constexpr auto step = 128u * 8u;
+        const auto blocks = (length & ~(step - 1));
+        for (auto i = 0u; i < blocks; i += step) {
+            XTEA_encrypt<128>(data + i, k);
+        }
+        data += blocks;
+        length -= blocks;
+    }
+#endif // __AVX512F__
+
+#ifdef __AVX__
+    {
+        constexpr auto step = 32u * 8u;
+        const auto blocks = (length & ~(step - 1));
+        for (auto i = 0u; i < blocks; i += step) {
+            XTEA_encrypt<32>(data + i, k);
+        }
+        data += blocks;
+        length -= blocks;
+    }
+#endif // __AVX__
+
+#if defined(__SSE__) || defined(__ARM_FEATURE_SIMD32)
+    {
+        constexpr auto step = 8u * 8u;
+        const auto blocks = (length & ~(step - 1));
+        for (auto i = 0u; i < blocks; i += step) {
+            XTEA_encrypt<8>(data + i, k);
+        }
+        data += blocks;
+        length -= blocks;
+    }
+#endif // __SSE__ || __ARM_FEATURE_SIMD32
+
+#ifdef __x86_64__
+    {
+        constexpr auto step = 2u * 8u;
+        const auto blocks = (length & ~(step - 1));
+        for (auto i = 0u; i < blocks; i += step) {
+            XTEA_encrypt<2>(data + i, k);
+        }
+        data += blocks;
+        length -= blocks;
+    }
+#endif // __x86_64__
```

Comment:
Since the only difference between `#ifdef` blocks is the block size, you could trim this down to:
```cpp
#ifdef __AVX512F__
    const auto block_size = 128u;
#endif // __AVX512F__

#ifdef __AVX__
    const auto block_size = 32u;
#endif // __AVX__

#if defined(__SSE__) || defined(__ARM_FEATURE_SIMD32)
    const auto block_size = 8u;
#endif // __SSE__ || __ARM_FEATURE_SIMD32

#ifdef __x86_64__
    const auto block_size = 2u;
#endif // __x86_64__

    constexpr auto step = block_size * 8u;
    const auto blocks = (length & ~(step - 1));
    for (auto i = 0u; i < blocks; i += step) {
        XTEA_encrypt<block_size>(data + i, k);
    }
    data += blocks;
    length -= blocks;
```
Other than that, the latest changes LGTM.

Reply:
Kinda, since each ifdef still enables the next one, as instruction sets include the previous ones. That is, `__AVX512F__` also enables `__AVX__` and `__SSE__`. I am still to test if the difference is significant, though.

Reply:
Well, doing as you suggest slows by less than 5%, which is bearable IMHO.

Reply:
Personally, I prefer speed over code reduction. I would rather keep that 5% in this case.

Reply:
5% is what new processors offer each year, so it seems like people care.

Reply:
Managed to make a cleaner code with some preprocessing and templates.